diff -urN a/include/ieee80211.h b/include/ieee80211.h
--- a/include/ieee80211.h	2026-04-23 09:55:41.216565159 +0200
+++ b/include/ieee80211.h	2026-04-23 09:45:14.261095931 +0200
@@ -1529,18 +1529,18 @@
 	(((Addr[2]) & 0xff) == 0xff) && (((Addr[3]) & 0xff) == 0xff) && (((Addr[4]) & 0xff) == 0xff) && \
 				     (((Addr[5]) & 0xff) == 0xff))
 #else
-extern __inline int is_multicast_mac_addr(const u8 *addr)
+static inline int is_multicast_mac_addr(const u8 *addr)
 {
 	return (addr[0] != 0xff) && (0x01 & addr[0]);
 }
 
-extern __inline int is_broadcast_mac_addr(const u8 *addr)
+static inline int is_broadcast_mac_addr(const u8 *addr)
 {
 	return ((addr[0] == 0xff) && (addr[1] == 0xff) && (addr[2] == 0xff) &&   \
 		(addr[3] == 0xff) && (addr[4] == 0xff) && (addr[5] == 0xff));
 }
 
-extern __inline int is_zero_mac_addr(const u8 *addr)
+static inline int is_zero_mac_addr(const u8 *addr)
 {
 	return ((addr[0] == 0x00) && (addr[1] == 0x00) && (addr[2] == 0x00) &&   \
 		(addr[3] == 0x00) && (addr[4] == 0x00) && (addr[5] == 0x00));
diff -urN a/include/osdep_service_linux.h b/include/osdep_service_linux.h
--- a/include/osdep_service_linux.h	2026-04-23 09:55:41.215402257 +0200
+++ b/include/osdep_service_linux.h	2026-04-23 09:32:24.635352298 +0200
@@ -50,6 +50,7 @@
 #include <linux/rtnetlink.h>
 #include <linux/delay.h>
 #include <linux/interrupt.h>	/* for struct tasklet_struct */
+#include <linux/timer.h>
 #include <linux/ip.h>
 #include <linux/kthread.h>
 #include <linux/list.h>
@@ -79,6 +80,16 @@
 /* Monitor mode */
 #include <net/ieee80211_radiotap.h>
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 17, 0))
+	#define rtw_from_timer(var, callback_timer, timer_fieldname) \
+		timer_container_of(var, callback_timer, timer_fieldname)
+	#define rtw_del_timer_sync timer_delete_sync
+#else
+	#define rtw_from_timer(var, callback_timer, timer_fieldname) \
+		from_timer(var, callback_timer, timer_fieldname)
+	#define rtw_del_timer_sync del_timer_sync
+#endif
+
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 24))
 	#include <linux/ieee80211.h>
 #endif
@@ -356,7 +367,7 @@
 #endif
 {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0))
-	_timer *ptimer = from_timer(ptimer, in_timer, timer);
+	_timer *ptimer = rtw_from_timer(ptimer, in_timer, timer);
 #else
 	_timer *ptimer = (_timer *)cntx;
 #endif
@@ -385,7 +396,7 @@
 
 __inline static void _cancel_timer(_timer *ptimer, u8 *bcancelled)
 {
-	*bcancelled = del_timer_sync(&ptimer->timer) == 1 ? 1 : 0;
+	*bcancelled = rtw_del_timer_sync(&ptimer->timer) == 1 ? 1 : 0;
 }
 
 static inline void _init_workitem(_workitem *pwork, void *pfunc, void *cntx)
diff -urN a/Makefile b/Makefile
--- a/Makefile	2026-04-23 09:55:41.215138678 +0200
+++ b/Makefile	2026-04-23 09:50:44.433506750 +0200
@@ -10,6 +10,8 @@
 EXTRA_CFLAGS += -Wno-unused
 EXTRA_CFLAGS += -Wno-vla -g
 EXTRA_CFLAGS += -Wno-implicit-fallthrough
+EXTRA_CFLAGS += -Wno-missing-prototypes
+EXTRA_CFLAGS += -Wno-missing-declarations
 
 GCC_VER_49 := $(shell echo `$(CC) -dumpversion | cut -f1-2 -d.` \>= 4.9 | bc )
 ifeq ($(GCC_VER_49),1)
@@ -75,6 +77,7 @@
 CONFIG_RTW_IPCAM_APPLICATION = n
 CONFIG_RTW_REPEATER_SON = n
 CONFIG_RTW_WIFI_HAL = n
+CONFIG_IOCTL_CFG80211 = n
 ########################## Debug ###########################
 CONFIG_RTW_DEBUG = y
 # default log level is _DRV_INFO_ = 4,
@@ -191,13 +194,16 @@
 			os_dep/linux/xmit_linux.o \
 			os_dep/linux/mlme_linux.o \
 			os_dep/linux/recv_linux.o \
-			os_dep/linux/ioctl_cfg80211.o \
-			os_dep/linux/rtw_cfgvendor.o \
-			os_dep/linux/wifi_regd.o \
 			os_dep/linux/rtw_android.o \
 			os_dep/linux/rtw_proc.o \
 			os_dep/linux/rtw_rhashtable.o
 
+ifeq ($(CONFIG_IOCTL_CFG80211), y)
+_OS_INTFS_FILES += os_dep/linux/ioctl_cfg80211.o \
+			os_dep/linux/rtw_cfgvendor.o \
+			os_dep/linux/wifi_regd.o
+endif
+
 ifeq ($(CONFIG_MP_INCLUDED), y)
 _OS_INTFS_FILES += os_dep/linux/ioctl_mp.o
 endif
@@ -1208,7 +1214,9 @@
 
 ifeq ($(CONFIG_PLATFORM_I386_PC), y)
 EXTRA_CFLAGS += -DCONFIG_LITTLE_ENDIAN
+ifeq ($(CONFIG_IOCTL_CFG80211), y)
 EXTRA_CFLAGS += -DCONFIG_IOCTL_CFG80211 -DRTW_USE_CFG80211_STA_EVENT
+endif
 SUBARCH := $(shell uname -m | sed -e "s/armv.*/arm/" | sed -e "s/aarch64/arm64/" | sed -e s/i.86/i386/)
 ARCH ?= $(SUBARCH)
 CROSS_COMPILE ?=
@@ -2146,6 +2154,9 @@
 include $(src)/rtl8821c.mk
 endif
 
+subdir-ccflags-y += $(EXTRA_CFLAGS)
+subdir-ccflags-y += -I$(src)/hal/phydm
+
 rtk_core :=	core/rtw_cmd.o \
 		core/rtw_security.o \
 		core/rtw_debug.o \
diff -urN a/os_dep/linux/os_intfs.c b/os_dep/linux/os_intfs.c
--- a/os_dep/linux/os_intfs.c	2026-04-23 09:55:41.216019215 +0200
+++ b/os_dep/linux/os_intfs.c	2026-04-23 09:40:24.293562803 +0200
@@ -1643,7 +1643,11 @@
 	u8 rtnl_lock_needed = rtw_rtnl_lock_needed(dvobj);
 
 #ifdef CONFIG_RTW_NAPI
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 17, 0))
+	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll);
+#else
 	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT);
+#endif
 #endif /* CONFIG_RTW_NAPI */
 
 #if defined(CONFIG_IOCTL_CFG80211)
diff -urN a/os_dep/linux/usb_intf.c b/os_dep/linux/usb_intf.c
--- a/os_dep/linux/usb_intf.c	2026-04-23 09:55:41.216264299 +0200
+++ b/os_dep/linux/usb_intf.c	2026-04-23 09:41:32.376664671 +0200
@@ -84,6 +84,13 @@
 	}
 }
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 17, 0))
+static void rtw_usb_shutdown(struct usb_interface *intf)
+{
+	rtw_dev_shutdown(&intf->dev);
+}
+#endif
+
 #if (LINUX_VERSION_CODE <= KERNEL_VERSION(2, 6, 23))
 /* Some useful macros to use to create struct usb_device_id */
 #define USB_DEVICE_ID_MATCH_VENDOR			 0x0001
@@ -281,7 +288,9 @@
 	.usbdrv.supports_autosuspend = 1,
 #endif
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 19))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 17, 0))
+	.usbdrv.shutdown = rtw_usb_shutdown,
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 19))
 	.usbdrv.drvwrap.driver.shutdown = rtw_dev_shutdown,
 #else
 	.usbdrv.driver.shutdown = rtw_dev_shutdown,
diff -urN a/os_dep/osdep_service.c b/os_dep/osdep_service.c
--- a/os_dep/osdep_service.c	2026-04-23 09:55:41.215649305 +0200
+++ b/os_dep/osdep_service.c	2026-04-23 09:40:13.839159053 +0200
@@ -17,6 +17,7 @@
 #define _OSDEP_SERVICE_C_
 
 #include <drv_types.h>
+#include <linux/random.h>
 
 #define RT_TAG	'1178'
 
@@ -2614,7 +2615,9 @@
 inline u32 rtw_random32(void)
 {
 #ifdef PLATFORM_LINUX
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 17, 0))
+	return get_random_u32();
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0))
 	return prandom_u32();
 #elif (LINUX_VERSION_CODE <= KERNEL_VERSION(2, 6, 18))
 	u32 random_int;
